### ISSUE TITLE
Add helper API on RRef to call functions on remote objects

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#31743]
### Problem

I frequently run into the situation where I have an `RRef` referencing a remote object, and I need to call a function on that remote object. Currently, we would need to first implement a wrapper function, and then call RPC on that function. For example:

```python
class ExampleClass:
    def some_func(self, arg1, arg2):
        pass

def call_some_func(rref, *args):
    rref.local_value().some_func(*args)

rref = remote("worker1", ExampleClass)
rpc_sync(rref.owner(), call_some_func, args=(1, 2))
```

Or more generically:

```python
def _run_remote_func(rref, func_name, *args, **kwargs):
    getattr(rref.local_value(), func_name)(*args, **kwargs)
```

### Proposal

It will be easier if `RRef` provides this feature natively. We cannot directly support `RRef.some_func(1, 2)`, as there could be name collisions between user functions and `RRef` API. And as discussed in #28882, it might be bad to trigger an implicit RPC here. 

What about something like the following?

```python
rref.remote_value().some_func(1, 2)
```

The API `remote_value()` pairs with the existing `local_value()` API, meaning that it is trying to access the value reference but remotely. It also avoids the name collision and implicit RPC problems above.  

In terms of implementation, we could let `RRef.remote_value()` return a special object `X` and also implement a utility function, e.g., `call_remote_value(rref, func_name, *args, **kwargs)`. In `X`'s `__getattr__` method, it partially applies the rref and the attribute name to `call_remote_value`, and returns the partially applied function. And `call_remote_value` uses RPC to invoke the `func_name` remotely.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528 @jjlilley

**Comment by mrshenli** [comment]
cc @agolynski @aazzolini @gqchen

**Comment by pritamdamania87** [comment]
Minor nit: Can we have `rref.remote().some_func(1, 2)` instead? Also, we could have name collisions for `call_remote_value` for the special object X as well right? Maybe we can implement the entire logic inside `__getattr__` and avoid having a utility function?

**Comment by mrshenli** [comment]
> Can we have rref.remote().some_func(1, 2) instead?

Yep, `remote` would also work, but not sure whether having two different functions of the same name (`rpc.remote()` and `rref.remote()`) would be confusing.

> Also, we could have name collisions for call_remote_value for the special object X as well right?

The special object X can be purely internal and only serve this single purpose. You mean the name collision could happen if the user overrides the `__getattr__` function on that special object X? I guess that's the only place where we would hit a collision.

> Maybe we can implement the entire logic inside `__getattr__` and avoid having a utility function?

You mean `RRef`'s `__getattr__`? That would be the best in terms of simplicity, and the only concern is API name collision on `RRef[T]` and `T`.

**Comment by pritamdamania87** [comment]
> You mean the name collision could happen if the user overrides the __getattr__ function on that special object X? I guess that's the only place where we would hit a collision.

No, I meant if the user defines their own function called `call_remote_value` and then call `rref.remote().call_remote_value(....)`.

> You mean RRef's __getattr__? That would be the best in terms of simplicity, and the only concern is API name collision on RRef[T] and T.

No, I meant instead of having a method called `call_remote_value` for the special object X, put the logic for `call_remote_value` inside `__getattr__` for the special object. This way you don't need a special function and avoid name collisions.

**Comment by pietern** [comment]
What about the different types of calls you could make, `remote`, `sync`, and `async`?

To disambiguate from these top level RPC functions, you could consider encoding this in the proxy object explicitly, and have `proxy_object_call_remote`, `proxy_object_call_sync`, and `proxy_object_call_async` (or variants that are less verbose).

**Comment by mrshenli** [comment]
> No, I meant if the user defines their own function called call_remote_value and then call `rref.remote().call_remote_value(....)`.

It should be fine I think. Because `rref.remote()` would return a special object X, which only has an overridden  `__getattr__` method. So when the app calls `X.call_remote_value(...)`, `X.__getattr__` will take the function name `"call_remote_value"` and pass it our own `call_remote_value` function as an arg. On the remote side, our own `call_remote_value` would invoke the user function on RRef by calling `getattr(rref.local_value(), "call_remote_value")`. This looks OK? 

> No, I meant instead of having a method called call_remote_value for the special object X, put the logic for call_remote_value inside __getattr__ for the special object. This way you don't need a special function and avoid name collisions.

Yep, we are talking about the same thing :). The difference is extracting that logical into a helper function (not on X) or not.

> What about the different types of calls you could make, remote, sync, and async?

Yep, that's indeed an issue. 

> To disambiguate from these top level RPC functions, you could consider encoding this in the proxy object explicitly, and have proxy_object_call_remote, proxy_object_call_sync, and proxy_object_call_async (or variants that are less verbose).

Do you mean sth like `rref.remote().user_func()`, `rref.rpc_async().user_func()`, and `rref.rpc_sync().user_func()`?

**Comment by pritamdamania87** [comment]
Just realized that one challenge here would be calling  something like `rref.remote().some_func(1, 2)` within a torchscript function. Torchscript doesn't support *args and **kwargs and as a result this could be a potential issue. 

cc @suo This is a good use case where supporting *args and **kwargs in torchscript would be useful.

**Comment by mrshenli** [comment]
closed by #36619
